# Pràctica aprenentatge automàtic

A partir de un data set donat (conjunt de dades), de jugadors del joc FIFA, hem de ser capaços de fer una predicció del valor econòmic del jugador segons les seves característiques. Intentar obtenir el % més alt d'encert possible.

Per a realitzar-ho, dividirem el problema en diverses parts:

1. Analitzar la base de dades per a saber quines columnes necessitem i quines no.

2. Eliminar columnes no necessàries.

 3. Observar els NaN's de la base de dades (significa que no hi ha informació).

4. Eliminar jugadors que no aportin informació i arreglar columnes amb NaN's.

5. Transformar totes les columnes a valors numèrics.

6. Convertir dades com per exemple de peus a centímetres (altura).

7. Una vegada arreglada tota la taula, entrenar amb el nostre dataset i obtenir la predicció.

Per a començar, importarem les llibreries que necessitarem:

In [1]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

## 1. Analitzar el data set

Assignem el nostre dataset a la variable df. Amb df.head() podem veure un cert nombre reduït de files i columnes.

In [2]:
df = pd.read_csv(os.path.join("..","in","fifa.csv"))
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]

Alterem les opcions per a que amb df.head() ens surtin un nombre de files reduït i totes les columnes.

In [3]:
pd.set_option('display.max_columns',100) #desplegar màxim de columnes
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

## 2. Eliminar columnes

Ara passem a analitzar el nostre data set per a identificar les columnes que no eliminarem, és a dir, no ens aporten informació que necessitem per a la predicció. Uns exemples d'aquestes serien les columnes ID ja que només fa d'identificador o Photo ja que només és la fotografia del jugador. Mitjançant la instrucció df.drop() eliminem dites columnes. Hi ha dues variants de drop() ja que amb la primera columna de "Unnamed:0" he tengut problemes i, per tant, l'he eliminat segons la seva posició de segons les columnes del dataset.

Les columnes: Photo, Flag i Club Logo les he eliminat ja que són només fotografies.

Les columnes: Unnamed 0:, ID, Name, Special (nombre de cartes especials d'aquell jugador), Real Face (si el joc té la seva cara real), Jersey Number (ndorsal que utilitza), Joined (quan va arribar al club), Loaned From ( si està cedit, desde quin club), Contract Valid Until (fins quan te contracte amb el club actual) les he eliminat ja que trob que no aporten informació rellevant.

In [4]:
df = df.drop(df.columns[[0]],axis='columns') #eliminar columna Unnamed:0
df = df.drop(columns=['ID','Name','Photo','Flag','Club Logo','Special',
                      'Real Face','Jersey Number','Joined','Loaned From','Contract Valid Until'])#eliminar la resta de columnes
df.head() #per a poder veure el resultat d'eliminar columnes i com queda la taula

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

        Work Rate   Body Type Position Height  Weight    LS    ST    RS    LW  \
0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2   
1       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3   
2    High/ Medium      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3   
3  Medium/ Medium        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   
4      High/ High      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3   

     LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   

    LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0   
4  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
3       13.0             21.0          50.0     13.0       18.0   21.0   
4       82.0             55.0          92.0     82.0       86.0   85.0   

   FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
0        94.0         87.0         96.0          91.0         86.0     91.0   
1        76.0         77.0         94.0          89.0         91.0     87.0   
2        87.0         78.0         95.0          94.0         90.0     96.0   
3        19.0         51.0         42.0          57.0         58.0     60.0   
4        83.0         91.0         91.0          78.0         76.0     79.0   

   Reactions  Balance  ShotPower  Jumping  Stamina  Strength  LongShots  \
0       95.0     95.0       85.0     68.0     72.0      59.0       94.0   
1       96.0     70.0       95.0     95.0     88.0      79.0       93.0   
2       94.0     84.0       80.0     61.0     81.0      49.0       82.0   
3       90.0     43.0       31.0     67.0     43.0      64.0       12.0   
4       91.0     77.0       91.0     63.0     90.0      75.0       91.0   

   Aggression  Interceptions  Positioning  Vision  Penalties  Composure  \
0        48.0           22.0         94.0    94.0       75.0       96.0   
1        63.0           29.0         95.0    82.0       85.0       95.0   
2        56.0           36.0         89.0    87.0       81.0       94.0   
3        38.0           30.0       

## 3. Observar quines columnes tenen NaN's

Una vegada eliminades les columnes, primer de tot analitzarem quines d'aquelles que hem deixat contenen NaN's.

In [5]:
df.columns[df.isna().any()].tolist() #llistat amb totes les columnes amb NaN's

['Club',
 'Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Body Type',
 'Position',
 'Height',
 'Weight',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

## 4. Eliminar jugadors i arreglar columnes amb NaN's

A partir d'aquí, primer mirem amb df.loc() les característiques dels jugadors que no tenen club. Com es pot observar alguns d'aquests tampoc tenen valors en les posicions LS, ST, ... . Aquests jugadors no tenen un Club que es trobi dins FIFA.

In [6]:
df.loc[df.Club != df.Club]

Age Nationality  Overall  Potential Club Value Wage Preferred Foot  \
452     24   Argentina       80         85  NaN    €0   €0          Right   
538     33      Sweden       80         80  NaN    €0   €0          Right   
568     26      Russia       79         81  NaN    €0   €0          Right   
677     29      Russia       79         79  NaN    €0   €0          Right   
874     29      Russia       78         78  NaN    €0   €0          Right   
...    ...         ...      ...        ...  ...   ...  ...            ...   
17197   21       India       55         64  NaN    €0   €0          Right   
17215   26     Finland       55         57  NaN    €0   €0          Right   
17339   23       India       54         63  NaN    €0   €0          Right   
17436   20       India       54         67  NaN    €0   €0          Right   
17539   21       India       53         62  NaN    €0   €0          Right   

       International Reputation  Weak Foot  Skill Moves       Work Rate  \
452                         2.0        4.0          4.0  Medium/ Medium   
538                         2.0        4.0          2.0    High/ Medium   
568                         1.0        3.0          1.0  Medium/ Medium   
677                         2.0        3.0          3.0      High/ High   
874                         2.0        3.0          3.0    High/ Medium   
...                         ...        ...          ...             ...   
17197                       1.0        2.0          1.0  Medium/ Medium   
17215                       1.0        3.0          2.0    Medium/ High   
17339                       1.0        3.0          2.0     Medium/ Low   
17436                       1.0        3.0          2.0  Medium/ Medium   
17539                       1.0        3.0          2.0    High/ Medium   

      Body Type Position Height  Weight    LS    ST    RS    LW    LF    CF  \
452      Normal       CM   5'11  165lbs  71+2  71+2  71+2  75+2  75+2  75+2   
538      Normal      LCB    6'4  185lbs  62+2  62+2  62+2  56+2  58+2  58+2   
568      Normal       GK    6'2  176lbs   NaN   NaN   NaN   NaN   NaN   NaN   
677        Lean       RB   5'10  154lbs  70+2  70+2  70+2  73+2  72+2  72+2   
874      Stocky       ST    6'5  201lbs  77+2  77+2  77+2  71+2  74+2  74+2   
...         ...      ...    ...     ...   ...   ...   ...   ...   ...   ...   
17197    Normal       GK    6'2  176lbs   NaN   NaN   NaN   NaN   NaN   NaN   
17215    Normal       RB    6'4  187lbs  46+2  46+2  46+2  45+2  44+2  44+2   
17339    Normal      NaN    5'9  143lbs   NaN   NaN   NaN   NaN   NaN   NaN   
17436    Normal      NaN    6'0  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   
17539      Lean      NaN    6'3  174lbs   NaN   NaN   NaN   NaN   NaN   NaN   

         RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
452    75+2  75+2  77+2  77+2  77+2  76+2  79+2  79+2  79+2  76+2  75+2  77+2   
538    58+2  56+2  58+2  58+2  58+2  57+2  64+2  64+2  64+2  57+2  68+2  74+2   
568     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
677    72+2  73+2  73+2  73+2  73+2  75+2  74+2  74+2  74+2  75+2  78+2  75+2   
874    74+2  71+2  71+2  71+2  71+2  71+2  66+2  66+2  66+2  71+2  52+2  52+2   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17197   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17215  44+2  45+2  43+2  43+2  43+2  46+2  43+2  43+2  43+2  46+2  51+2  49+2   
17339   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17436   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17539   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

        CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
452    77+2  77+2  75+2  74+2  72+2  72+2  72+2  74+2      76.0       55.0   
538    74+2  74+2  68+2  70+2  79+2  79+2  79+2  70+2      49.0       51.0   
568     NaN   NaN   NaN  

Així que eliminem tots els jugadors que no tenen club amb dropna() que elimina totes les files que tinguin NaN a la columna del club.

In [7]:
df = df.dropna(subset = ['Club'])
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
3          Right                       4.0        3.0          1.0   
4          Right                       4.0        5.0          4.0   

        Work Rate   Body Type Position Height  Weight    LS    ST    RS    LW  \
0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2   
1       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3   
2    High/ Medium      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3   
3  Medium/ Medium        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   
4      High/ High      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3   

     LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   

    LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0   
4  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
3       13.0             21.0          50.0     13.0       18.0   21.0   
4       82.0             55.0          92.0     82.0       86.0   85.0   

   FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
0        94.0         87.0         96.0          91.0         86.0     91.0   
1        76.0         77.0         94.0          89.0         91.0     87.0   
2        87.0         78.0         95.0          94.0         90.0     96.0   
3        19.0         51.0         42.0          57.0         58.0     60.0   
4        83.0         91.0         91.0          78.0         76.0     79.0   

   Reactions  Balance  ShotPower  Jumping  Stamina  Strength  LongShots  \
0       95.0     95.0       85.0     68.0     72.0      59.0       94.0   
1       96.0     70.0       95.0     95.0     88.0      79.0       93.0   
2       94.0     84.0       80.0     61.0     81.0      49.0       82.0   
3       90.0     43.0       31.0     67.0     43.0      64.0       12.0   
4       91.0     77.0       91.0     63.0     90.0      75.0       91.0   

   Aggression  Interceptions  Positioning  Vision  Penalties  Composure  \
0        48.0           22.0         94.0    94.0       75.0       96.0   
1        63.0           29.0         95.0    82.0       85.0       95.0   
2        56.0           36.0         89.0    87.0       81.0       94.0   
3        38.0           30.0       

Com es pot observar, ja no hi ha jugadors que no tinguin club.

In [8]:
df.loc[df.Club != df.Club]

Empty DataFrame
Columns: [Age, Nationality, Overall, Potential, Club, Value, Wage, Preferred Foot, International Reputation, Weak Foot, Skill Moves, Work Rate, Body Type, Position, Height, Weight, LS, ST, RS, LW, LF, CF, RF, RW, LAM, CAM, RAM, LM, LCM, CM, RCM, RM, LWB, LDM, CDM, RDM, RWB, LB, LCB, CB, RCB, RB, Crossing, Finishing, HeadingAccuracy, ShortPassing, Volleys, Dribbling, Curve, FKAccuracy, LongPassing, BallControl, Acceleration, SprintSpeed, Agility, Reactions, Balance, ShotPower, Jumping, Stamina, Strength, LongShots, Aggression, Interceptions, Positioning, Vision, Penalties, Composure, Marking, StandingTackle, SlidingTackle, GKDiving, GKHandling, GKKicking, GKPositioning, GKReflexes, Release Clause]
Index: []

Comprovem els jugadors que no tenen valors a LS. Mirant la relació que hi ha entre les columnes Position i LS podem veure que bastants dels jugadors que no tenen valors a LS, són porters. Per tant, els eliminarem de la nostra base de dades.

In [9]:
df.loc[df.LS != df.LS]

Age Nationality  Overall  Potential               Club   Value   Wage  \
3       27       Spain       91         93  Manchester United    €72M  €260K   
9       25    Slovenia       90         93    Atlético Madrid    €68M   €94K   
18      26     Germany       89         92       FC Barcelona    €58M  €240K   
19      26     Belgium       89         90        Real Madrid  €53.5M  €240K   
22      32     Germany       89         89  FC Bayern München    €38M  €130K   
...    ...         ...      ...        ...                ...     ...    ...   
18178   18      Sweden       48         65         Dalkurd FF    €50K    €1K   
18180   22    Scotland       48         58   St. Johnstone FC    €40K    €1K   
18183   44     England       48         48   Cambridge United      €0    €1K   
18194   18       Italy       47         65              Lecce    €50K    €1K   
18198   18     England       47         70      Burton Albion    €60K    €1K   

      Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
3              Right                       4.0        3.0          1.0   
9              Right                       3.0        3.0          1.0   
18             Right                       3.0        4.0          1.0   
19              Left                       4.0        2.0          1.0   
22             Right                       5.0        4.0          1.0   
...              ...                       ...        ...          ...   
18178          Right                       1.0        2.0          1.0   
18180          Right                       1.0        2.0          1.0   
18183          Right                       1.0        2.0          1.0   
18194          Right                       1.0        3.0          1.0   
18198          Right                       1.0        2.0          1.0   

            Work Rate Body Type Position Height  Weight   LS   ST   RS   LW  \
3      Medium/ Medium      Lean       GK    6'4  168lbs  NaN  NaN  NaN  NaN   
9      Medium/ Medium    Normal       GK    6'2  192lbs  NaN  NaN  NaN  NaN   
18     Medium/ Medium    Normal       GK    6'2  187lbs  NaN  NaN  NaN  NaN   
19     Medium/ Medium  Courtois       GK    6'6  212lbs  NaN  NaN  NaN  NaN   
22     Medium/ Medium    Normal       GK    6'4  203lbs  NaN  NaN  NaN  NaN   
...               ...       ...      ...    ...     ...  ...  ...  ...  ...   
18178  Medium/ Medium      Lean       GK    6'0  176lbs  NaN  NaN  NaN  NaN   
18180  Medium/ Medium      Lean       GK    6'1  172lbs  NaN  NaN  NaN  NaN   
18183  Medium/ Medium    Normal       GK    6'2  190lbs  NaN  NaN  NaN  NaN   
18194  Medium/ Medium    Normal       GK    6'3  187lbs  NaN  NaN  NaN  NaN   
18198  Medium/ Medium      Lean       GK   5'11  154lbs  NaN  NaN  NaN  NaN   

        LF   CF   RF   RW  LAM  CAM  RAM   LM  LCM   CM  RCM   RM  LWB  LDM  \
3      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
19     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
22     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
18178  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18180  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18183  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18194  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18198  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

       CDM  RDM  RWB   LB  LCB   CB  RCB   RB  Crossing  Finishing  \
3      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      17.0       13.0   
9      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      13.0       11.0   
18     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN    

In [10]:
df.loc[df.Position =='GK']

Age Nationality  Overall  Potential               Club   Value   Wage  \
3       27       Spain       91         93  Manchester United    €72M  €260K   
9       25    Slovenia       90         93    Atlético Madrid    €68M   €94K   
18      26     Germany       89         92       FC Barcelona    €58M  €240K   
19      26     Belgium       89         90        Real Madrid  €53.5M  €240K   
22      32     Germany       89         89  FC Bayern München    €38M  €130K   
...    ...         ...      ...        ...                ...     ...    ...   
18178   18      Sweden       48         65         Dalkurd FF    €50K    €1K   
18180   22    Scotland       48         58   St. Johnstone FC    €40K    €1K   
18183   44     England       48         48   Cambridge United      €0    €1K   
18194   18       Italy       47         65              Lecce    €50K    €1K   
18198   18     England       47         70      Burton Albion    €60K    €1K   

      Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
3              Right                       4.0        3.0          1.0   
9              Right                       3.0        3.0          1.0   
18             Right                       3.0        4.0          1.0   
19              Left                       4.0        2.0          1.0   
22             Right                       5.0        4.0          1.0   
...              ...                       ...        ...          ...   
18178          Right                       1.0        2.0          1.0   
18180          Right                       1.0        2.0          1.0   
18183          Right                       1.0        2.0          1.0   
18194          Right                       1.0        3.0          1.0   
18198          Right                       1.0        2.0          1.0   

            Work Rate Body Type Position Height  Weight   LS   ST   RS   LW  \
3      Medium/ Medium      Lean       GK    6'4  168lbs  NaN  NaN  NaN  NaN   
9      Medium/ Medium    Normal       GK    6'2  192lbs  NaN  NaN  NaN  NaN   
18     Medium/ Medium    Normal       GK    6'2  187lbs  NaN  NaN  NaN  NaN   
19     Medium/ Medium  Courtois       GK    6'6  212lbs  NaN  NaN  NaN  NaN   
22     Medium/ Medium    Normal       GK    6'4  203lbs  NaN  NaN  NaN  NaN   
...               ...       ...      ...    ...     ...  ...  ...  ...  ...   
18178  Medium/ Medium      Lean       GK    6'0  176lbs  NaN  NaN  NaN  NaN   
18180  Medium/ Medium      Lean       GK    6'1  172lbs  NaN  NaN  NaN  NaN   
18183  Medium/ Medium    Normal       GK    6'2  190lbs  NaN  NaN  NaN  NaN   
18194  Medium/ Medium    Normal       GK    6'3  187lbs  NaN  NaN  NaN  NaN   
18198  Medium/ Medium      Lean       GK   5'11  154lbs  NaN  NaN  NaN  NaN   

        LF   CF   RF   RW  LAM  CAM  RAM   LM  LCM   CM  RCM   RM  LWB  LDM  \
3      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
9      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
19     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
22     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
18178  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18180  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18183  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18194  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
18198  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

       CDM  RDM  RWB   LB  LCB   CB  RCB   RB  Crossing  Finishing  \
3      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      17.0       13.0   
9      NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN      13.0       11.0   
18     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN    

Com veim, el nombre de jugadors que són porters és molt reduït respecte al conjunt de dades. Si els eliminem no afectarà en gran mesura al resultat obtingut. No hi haurà casi diferència entre un resultat tinguent en compte els portes que un sense tenir-los en compte 

Eliminació dels jugadors amb posició igual a GK, és a dir, que siguin porters.

In [11]:
df = df.drop(df[df['Position']=='GK'].index)

In [12]:
df.columns[df.isna().any()].tolist() #miram les columnes que encara tenen NaN's

['Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Body Type',
 'Position',
 'Height',
 'Weight',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

In [13]:
df.loc[df.Position != df.Position].head() #per a veure els jugadors que no tenen posició assignada

Age Nationality  Overall  Potential                    Club  Value  \
13236   33    Scotland       62         62                Rochdale  €120K   
13237   29   Nicaragua       62         62         Boyacá Chicó FC  €300K   
13238   35     England       62         62            Notts County  €140K   
13239   20       Italy       62         72                 Brescia  €425K   
13240   24     England       62         66  Hamilton Academical FC  €400K   

      Wage Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
13236  €1K            NaN                       NaN        NaN          NaN   
13237  €1K            NaN                       NaN        NaN          NaN   
13238  €3K            NaN                       NaN        NaN          NaN   
13239  €1K            NaN                       NaN        NaN          NaN   
13240  €1K            NaN                       NaN        NaN          NaN   

      Work Rate Body Type Position Height Weight   LS   ST   RS   LW   LF  \
13236       NaN       NaN      NaN    NaN    NaN  NaN  NaN  NaN  NaN  NaN   
13237       NaN       NaN      NaN    NaN    NaN  NaN  NaN  NaN  NaN  NaN   
13238       NaN       NaN      NaN    NaN    NaN  NaN  NaN  NaN  NaN  NaN   
13239       NaN       NaN      NaN    NaN    NaN  NaN  NaN  NaN  NaN  NaN   
13240       NaN       NaN      NaN    NaN    NaN  NaN  NaN  NaN  NaN  NaN   

        CF   RF   RW  LAM  CAM  RAM   LM  LCM   CM  RCM   RM  LWB  LDM  CDM  \
13236  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
13237  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
13238  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
13239  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
13240  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

       RDM  RWB   LB  LCB   CB  RCB   RB  Crossing  Finishing  \
13236  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN        NaN   
13237  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN        NaN   
13238  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN        NaN   
13239  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN        NaN   
13240  NaN  NaN  NaN  NaN  NaN  NaN  NaN       NaN        NaN   

       HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
13236              NaN           NaN      NaN        NaN    NaN         NaN   
13237              NaN           NaN      NaN        NaN    NaN         NaN   
13238              NaN           NaN      NaN        NaN    NaN         NaN   
13239              NaN           NaN      NaN        NaN    NaN         NaN   
13240              NaN           NaN      NaN        NaN    NaN         NaN   

       LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
13236          NaN          NaN           NaN          NaN      NaN   
13237          NaN          NaN           NaN          NaN      NaN   
13238          NaN          NaN           NaN          NaN      NaN   
13239          NaN          NaN           NaN          NaN      NaN   
13240          NaN          NaN           NaN          NaN      NaN   

       Reactions  Balance  ShotPower  Jumping  Stamina  Strength  LongShots  \
13236        NaN      NaN        NaN      NaN      NaN       NaN        NaN   
13237        NaN      NaN        NaN      NaN      NaN       NaN        NaN   
13238        NaN      NaN        NaN      NaN      NaN       NaN        NaN   
13239        NaN      NaN        NaN      NaN      NaN       NaN        NaN   
13240        NaN      NaN        NaN      NaN      NaN       NaN        NaN   

       Aggression  Interceptions  Positioning  Vision  Penalties  Composure  \
13236         NaN            NaN          NaN     NaN        NaN        NaN   
13237         NaN            NaN          NaN     NaN        NaN        NaN   
13238         NaN            NaN          NaN     NaN        NaN        NaN   
13239         NaN            NaN          NaN  

A continuació, també suprimim els jugadors que no tinguin posició assignada. La raó és que, aquests, són els que també tenen les altres columnes com Preferred Foot o Weak Foot amb NaN. És a dir, no ens aporten molta informació. FIFA no conta amb casi res d'informació d'aquests jugadors. 

In [14]:
df = df.dropna(subset = ['Position'])
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   
5   27     Belgium       91         91              Chelsea     €93M  €340K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
4          Right                       4.0        5.0          4.0   
5          Right                       4.0        4.0          4.0   

        Work Rate   Body Type Position Height  Weight    LS    ST    RS    LW  \
0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2   
1       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3   
2    High/ Medium      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3   
4      High/ High      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3   
5    High/ Medium      Normal       LF    5'8  163lbs  83+3  83+3  83+3  89+3   

     LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
4  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   
5  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3   

    LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0   
4  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0   
5  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
4       82.0             55.0          92.0     82.0       86.0   85.0   
5       84.0             61.0          89.0     80.0       95.0   83.0   

   FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
0        94.0         87.0         96.0          91.0         86.0     91.0   
1        76.0         77.0         94.0          89.0         91.0     87.0   
2        87.0         78.0         95.0          94.0         90.0     96.0   
4        83.0         91.0         91.0          78.0         76.0     79.0   
5        79.0         83.0         94.0          94.0         88.0     95.0   

   Reactions  Balance  ShotPower  Jumping  Stamina  Strength  LongShots  \
0       95.0     95.0       85.0     68.0     72.0      59.0       94.0   
1       96.0     70.0       95.0     95.0     88.0      79.0       93.0   
2       94.0     84.0       80.0     61.0     81.0      49.0       82.0   
4       91.0     77.0       91.0     63.0     90.0      75.0       91.0   
5       90.0     94.0       82.0     56.0     83.0      66.0       80.0   

   Aggression  Interceptions  Positioning  Vision  Penalties  Composure  \
0        48.0           22.0         94.0    94.0       75.0       96.0   
1        63.0           29.0         95.0    82.0       85.0       95.0   
2        56.0           36.0         89.0    87.0       81.0       94.0   
4        76.0           61.0       

Una vegada eliminats els jugadors amb posicions sense valors, podem veure que només ens queda la columna de Release Clause.
Problema: Els jugadors amb NaN's a la Release Clause no tenim perquè obligatòriament eliminar-los. Solució: suposarem que la seva Release Clause val 0.

In [15]:
df.columns[df.isna().any()].tolist()

['Release Clause']

In [16]:
df = df.fillna('0€') #afegim el signe € per a que no sorgeixin problemes amb un algoritme realitzat més endavant.
df.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   
5   27     Belgium       91         91              Chelsea     €93M  €340K   

  Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0           Left                       5.0        4.0          4.0   
1          Right                       5.0        4.0          5.0   
2          Right                       5.0        5.0          5.0   
4          Right                       4.0        5.0          4.0   
5          Right                       4.0        4.0          4.0   

        Work Rate   Body Type Position Height  Weight    LS    ST    RS    LW  \
0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2   
1       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3   
2    High/ Medium      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3   
4      High/ High      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3   
5    High/ Medium      Normal       LF    5'8  163lbs  83+3  83+3  83+3  89+3   

     LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
4  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   
5  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3   

    LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0   
4  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0   
5  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
4       82.0             55.0          92.0     82.0       86.0   85.0   
5       84.0             61.0          89.0     80.0       95.0   83.0   

   FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
0        94.0         87.0         96.0          91.0         86.0     91.0   
1        76.0         77.0         94.0          89.0         91.0     87.0   
2        87.0         78.0         95.0          94.0         90.0     96.0   
4        83.0         91.0         91.0          78.0         76.0     79.0   
5        79.0         83.0         94.0          94.0         88.0     95.0   

   Reactions  Balance  ShotPower  Jumping  Stamina  Strength  LongShots  \
0       95.0     95.0       85.0     68.0     72.0      59.0       94.0   
1       96.0     70.0       95.0     95.0     88.0      79.0       93.0   
2       94.0     84.0       80.0     61.0     81.0      49.0       82.0   
4       91.0     77.0       91.0     63.0     90.0      75.0       91.0   
5       90.0     94.0       82.0     56.0     83.0      66.0       80.0   

   Aggression  Interceptions  Positioning  Vision  Penalties  Composure  \
0        48.0           22.0         94.0    94.0       75.0       96.0   
1        63.0           29.0         95.0    82.0       85.0       95.0   
2        56.0           36.0         89.0    87.0       81.0       94.0   
4        76.0           61.0       

Realitzem la comprovació de que ja no queden NaN's a la taula.

In [17]:
df.columns[df.isna().any()].tolist()

[]

## 5. Adaptar columnes per a poder fer feina amb elles

Convertir a valors numèrics les columnes com Club o Preferred Foot ja que són Strings i necessitem que siguin valors per a tal de fer feina amb ells. Això ho realitzarem de tal manera que:
Per al Club construirem una columna per a cada variable que hi ha. Així se li assignarà un 0 si no pertany a aquest Club o un 1 si, sí que hi pertany.

Realitzarem la conversió per a diverses columnes: Club, Preferred Foot, Nationality, Body Type, Work Rate i Position.

In [18]:
clb = df.pop("Club") #sempre el mateix esquema, només canvien les columnes a les quals ho apliquem. Aquesta és per al "Club"

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)

In [19]:
nat = df.pop("Nationality") #"Nationality"

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(nat, prefix='nat').reset_index(drop=True)], axis=1, sort=False)

In [20]:
pf = df.pop("Preferred Foot") #"Preferred Foot"

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pf, prefix='pf').reset_index(drop=True)], axis=1, sort=False)

In [21]:
bt = df.pop("Body Type") #"Body Type"

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(bt, prefix='bt').reset_index(drop=True)], axis=1, sort=False)

In [22]:
wr = df.pop("Work Rate") #"Work Rate"

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(wr, prefix='wr').reset_index(drop=True)], axis=1, sort=False)

In [23]:
pos = df.pop("Position") #"Position"

df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pos, prefix='pos').reset_index(drop=True)], axis=1, sort=False)
df.head()

Age  Overall  Potential    Value   Wage  International Reputation  \
0   31       94         94  €110.5M  €565K                       5.0   
1   33       94         94     €77M  €405K                       5.0   
2   26       92         93  €118.5M  €290K                       5.0   
3   27       91         92    €102M  €355K                       4.0   
4   27       91         91     €93M  €340K                       4.0   

   Weak Foot  Skill Moves Height  Weight    LS    ST    RS    LW    LF    CF  \
0        4.0          4.0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1        4.0          5.0    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2        5.0          5.0    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        5.0          4.0   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   
4        4.0          4.0    5'8  163lbs  83+3  83+3  83+3  89+3  88+3  88+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   
4  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3  66+3  63+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   
4  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0       84.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             55.0          92.0     82.0       86.0   85.0        83.0   
4             61.0          89.0     80.0       95.0   83.0        79.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         91.0         91.0          78.0         76.0     79.0       91.0   
4         83.0         94.0          94.0         88.0     95.0       90.0   

   ...  nat_Venezuela  nat_Wales  nat_Zambia  nat_Zimbabwe  pf_Left  pf_Right  \
0  ...              0          0           0             0        1         0   
1  ...              0          0           0             0        0         1   
2  ...              0          0           0             0        0         1   
3  ...              0          0           0             0        0         1   
4  ...              0          0           0             0        0         1   

   bt_Akinfenwa  bt_C. Ronaldo  bt_Lean  bt_Messi  bt_Neymar  bt_Normal  \
0             0              0        0         1          0          0   
1             0              1        0         0          0          0   
2             0              0        0         0          1          0   
3             0              0        0         0          0          1   
4             0              0        0         0          0          1   

   bt_PLAYER_BODY_TYPE_25  bt_Shaqiri  bt_Stocky  wr_High/ High  wr_High/ Low  \
0                       0           0          0              0             0   
1                       0           0          0              0             1   
2                       0           0          0              0             0   

## 6. Convertir dades

Com es pot veure, el nombre de columnes ha augmentat considerablement però d'aquesta forma ja podem fer feina amb dites columnes. A continuació, toca arreglar les columnes que contenen lletres com Weight o Value (eliminar les lletres). També, cal realitzar una conversió a unitats de mesures que coneguem i a dades del tipus int o float. Height, està mesurat en peus i ho passarem a centímetres.

In [24]:
def value_to_float(x): #algoritme per aplicar a "Value", "Wage" i "Release Clause"
    """
    De K i M a float
    
    """
    x = x.replace('€', '') #eliminem el signe €
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x: #si conté una K
        if len(x) > 1:
            ret_val = float(x.replace('K', '')) #eliminem la K
        ret_val = ret_val *1000 #multipliquem per 1000 ja que la K simbolitzava aquest valor
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', '')) #eliminem la M
        ret_val = ret_val * 1000000.0 #multipliquem per 1000000 ja que la M simbolitzava aquest valor
    return ret_val #retornem el valor obtingut

In [25]:
df["Value"] = df["Value"].apply(value_to_float) #aplicar l'algoritme a la columna "Value"
df.head()

Age  Overall  Potential        Value   Wage  International Reputation  \
0   31       94         94  110500000.0  €565K                       5.0   
1   33       94         94   77000000.0  €405K                       5.0   
2   26       92         93  118500000.0  €290K                       5.0   
3   27       91         92  102000000.0  €355K                       4.0   
4   27       91         91   93000000.0  €340K                       4.0   

   Weak Foot  Skill Moves Height  Weight    LS    ST    RS    LW    LF    CF  \
0        4.0          4.0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1        4.0          5.0    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2        5.0          5.0    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        5.0          4.0   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   
4        4.0          4.0    5'8  163lbs  83+3  83+3  83+3  89+3  88+3  88+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   
4  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3  66+3  63+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   
4  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0       84.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             55.0          92.0     82.0       86.0   85.0        83.0   
4             61.0          89.0     80.0       95.0   83.0        79.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         91.0         91.0          78.0         76.0     79.0       91.0   
4         83.0         94.0          94.0         88.0     95.0       90.0   

   ...  nat_Venezuela  nat_Wales  nat_Zambia  nat_Zimbabwe  pf_Left  pf_Right  \
0  ...              0          0           0             0        1         0   
1  ...              0          0           0             0        0         1   
2  ...              0          0           0             0        0         1   
3  ...              0          0           0             0        0         1   
4  ...              0          0           0             0        0         1   

   bt_Akinfenwa  bt_C. Ronaldo  bt_Lean  bt_Messi  bt_Neymar  bt_Normal  \
0             0              0        0         1          0          0   
1             0              1        0         0          0          0   
2             0              0        0         0          1          0   
3             0              0        0         0          0          1   
4             0              0        0         0          0          1   

   bt_PLAYER_BODY_TYPE_25  bt_Shaqiri  bt_Stocky  wr_High/ High  wr_High/ Low  \
0                       0           0          0              0             0   
1                       0           0          0              0             1   
2                       0           0          0        

In [26]:
df["Wage"] = df["Wage"].apply(value_to_float) #aplicar l'algoritme a la columna "Wage"
df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         92  102000000.0  355000.0                       4.0   
4   27       91         91   93000000.0  340000.0                       4.0   

   Weak Foot  Skill Moves Height  Weight    LS    ST    RS    LW    LF    CF  \
0        4.0          4.0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1        4.0          5.0    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2        5.0          5.0    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        5.0          4.0   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   
4        4.0          4.0    5'8  163lbs  83+3  83+3  83+3  89+3  88+3  88+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   
4  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3  66+3  63+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   
4  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0       84.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             55.0          92.0     82.0       86.0   85.0        83.0   
4             61.0          89.0     80.0       95.0   83.0        79.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         91.0         91.0          78.0         76.0     79.0       91.0   
4         83.0         94.0          94.0         88.0     95.0       90.0   

   ...  nat_Venezuela  nat_Wales  nat_Zambia  nat_Zimbabwe  pf_Left  pf_Right  \
0  ...              0          0           0             0        1         0   
1  ...              0          0           0             0        0         1   
2  ...              0          0           0             0        0         1   
3  ...              0          0           0             0        0         1   
4  ...              0          0           0             0        0         1   

   bt_Akinfenwa  bt_C. Ronaldo  bt_Lean  bt_Messi  bt_Neymar  bt_Normal  \
0             0              0        0         1          0          0   
1             0              1        0         0          0          0   
2             0              0        0         0          1          0   
3             0              0        0         0          0          1   
4             0              0        0         0          0          1   

   bt_PLAYER_BODY_TYPE_25  bt_Shaqiri  bt_Stocky  wr_High/ High  wr_High/ Low  \
0                       0           0          0              0             0   
1                       0           0          0              0             1   
2                       0           0 

In [27]:
df["Release Clause"] = df["Release Clause"].apply(value_to_float) #aplicar l'algoritme a la columna "Release Clause"
df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         92  102000000.0  355000.0                       4.0   
4   27       91         91   93000000.0  340000.0                       4.0   

   Weak Foot  Skill Moves Height  Weight    LS    ST    RS    LW    LF    CF  \
0        4.0          4.0    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1        4.0          5.0    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2        5.0          5.0    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3        5.0          4.0   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   
4        4.0          4.0    5'8  163lbs  83+3  83+3  83+3  89+3  88+3  88+3   

     RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+3   
3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3  77+3   
4  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3  66+3  63+3   

    CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   
4  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0       84.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             55.0          92.0     82.0       86.0   85.0        83.0   
4             61.0          89.0     80.0       95.0   83.0        79.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         91.0         91.0          78.0         76.0     79.0       91.0   
4         83.0         94.0          94.0         88.0     95.0       90.0   

   ...  nat_Venezuela  nat_Wales  nat_Zambia  nat_Zimbabwe  pf_Left  pf_Right  \
0  ...              0          0           0             0        1         0   
1  ...              0          0           0             0        0         1   
2  ...              0          0           0             0        0         1   
3  ...              0          0           0             0        0         1   
4  ...              0          0           0             0        0         1   

   bt_Akinfenwa  bt_C. Ronaldo  bt_Lean  bt_Messi  bt_Neymar  bt_Normal  \
0             0              0        0         1          0          0   
1             0              1        0         0          0          0   
2             0              0        0         0          1          0   
3             0              0        0         0          0          1   
4             0              0        0         0          0          1   

   bt_PLAYER_BODY_TYPE_25  bt_Shaqiri  bt_Stocky  wr_High/ High  wr_High/ Low  \
0                       0           0          0              0             0   
1                       0           0          0              0             1   
2                       0           0 

Arreglar la columna de Weight amb el següent algoritme. Convertir de lliures a Kilograms. Eliminar les lletres lbs.

In [28]:
def value_to_kg(x):
    """
    Passar de lliures a Kilograms
    """
    x = float(x.replace('lbs','')) #eliminem lbs i convertim el valor a un de tipus float
    ret_val = x/2.2046; #fórmula de conversió, dividir les lliures entre 2.2046
    return ret_val #retornem el valor obtingut

In [29]:
df["Weight"] = df["Weight"].apply(value_to_kg) #aplicar l'algoritme a la columna "Weight"
df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         92  102000000.0  355000.0                       4.0   
4   27       91         91   93000000.0  340000.0                       4.0   

   Weak Foot  Skill Moves Height     Weight    LS    ST    RS    LW    LF  \
0        4.0          4.0    5'7  72.121927  88+2  88+2  88+2  92+2  93+2   
1        4.0          5.0    6'2  83.008255  91+3  91+3  91+3  89+3  90+3   
2        5.0          5.0    5'9  68.039554  84+3  84+3  84+3  89+3  89+3   
3        5.0          4.0   5'11  69.853942  82+3  82+3  82+3  87+3  87+3   
4        4.0          4.0    5'8  73.936315  83+3  83+3  83+3  89+3  88+3   

     CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB  \
0  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2   
1  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3   
3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3  77+3   
4  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3  66+3   

    LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  Finishing  \
0  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0       95.0   
1  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0       94.0   
2  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0       87.0   
3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0       82.0   
4  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0       84.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             55.0          92.0     82.0       86.0   85.0        83.0   
4             61.0          89.0     80.0       95.0   83.0        79.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         91.0         91.0          78.0         76.0     79.0       91.0   
4         83.0         94.0          94.0         88.0     95.0       90.0   

   ...  nat_Venezuela  nat_Wales  nat_Zambia  nat_Zimbabwe  pf_Left  pf_Right  \
0  ...              0          0           0             0        1         0   
1  ...              0          0           0             0        0         1   
2  ...              0          0           0             0        0         1   
3  ...              0          0           0             0        0         1   
4  ...              0          0           0             0        0         1   

   bt_Akinfenwa  bt_C. Ronaldo  bt_Lean  bt_Messi  bt_Neymar  bt_Normal  \
0             0              0        0         1          0          0   
1             0              1        0         0          0          0   
2             0              0        0         0          1          0   
3             0              0        0         0          0          1   
4             0              0        0         0          0          1   

   bt_PLAYER_BODY_TYPE_25  bt_Shaqiri  bt_Stocky  wr_High/ High  wr_High/ Low  \
0                       0           0          0              0             0   
1                       0           0          0              0             1   
2                   

Convertir de peus a centímetres.

In [30]:
def value_to_cm(x):
    """
    Passar de peus a centímetres
    """
    x = float(x.replace("'",".")) #eliminar la ' per a un . per a poder fer feina. Convertir a un valor float
    ret_val = x/0.0328084; #dividir per a obtenir centímetres
    return ret_val #retornar el resultat obtingut

In [31]:
df["Height"] = df["Height"].apply(value_to_cm) #aplicar l'algoritme a la columna "Height"
df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         92  102000000.0  355000.0                       4.0   
4   27       91         91   93000000.0  340000.0                       4.0   

   Weak Foot  Skill Moves      Height     Weight    LS    ST    RS    LW  \
0        4.0          4.0  173.735994  72.121927  88+2  88+2  88+2  92+2   
1        4.0          5.0  188.975994  83.008255  91+3  91+3  91+3  89+3   
2        5.0          5.0  179.831994  68.039554  84+3  84+3  84+3  89+3   
3        5.0          4.0  155.752795  69.853942  82+3  82+3  82+3  87+3   
4        4.0          4.0  176.783994  73.936315  83+3  83+3  83+3  89+3   

     LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   
4  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3  82+3  82+3  82+3  89+3   

    LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0   
3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0   
4  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3  49+3  60+3      81.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
3       82.0             55.0          92.0     82.0       86.0   85.0   
4       84.0             61.0          89.0     80.0       95.0   83.0   

   FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
0        94.0         87.0         96.0          91.0         86.0     91.0   
1        76.0         77.0         94.0          89.0         91.0     87.0   
2        87.0         78.0         95.0          94.0         90.0     96.0   
3        83.0         91.0         91.0          78.0         76.0     79.0   
4        79.0         83.0         94.0          94.0         88.0     95.0   

   Reactions  ...  nat_Venezuela  nat_Wales  nat_Zambia  nat_Zimbabwe  \
0       95.0  ...              0          0           0             0   
1       96.0  ...              0          0           0             0   
2       94.0  ...              0          0           0             0   
3       91.0  ...              0          0           0             0   
4       90.0  ...              0          0           0             0   

   pf_Left  pf_Right  bt_Akinfenwa  bt_C. Ronaldo  bt_Lean  bt_Messi  \
0        1         0             0              0        0         1   
1        0         1             0              1        0         0   
2        0         1             0              0        0         0   
3        0         1             0              0        0         0   
4        0         1             0              0        0         0   

   bt_Neymar  bt_Normal  bt_PLAYER_BODY_TYPE_25  bt_Shaqiri  bt_Stocky  \
0          0          0                       0           0          0   
1          0          0                       0           0          0   
2          1          0                       0           0          0   
3          0          1                       0           0          

Falta arreglar els valors de diverses columnes com LS,ST,...
Ho realitzarem mitjançant l'aplicació de la instrucció partition().

In [32]:
def value_to_int(x): #algoritme
    """
    Pasar de variable de tipus no enter a enter
    """
    primerval = x.partition("+")[0] #seleccionar els dígits de abans del + i guardar-los en una variable
    primerval_entero = int(primerval) #variable de tipus enter
    segundoval = x.partition("+")[2] #seleccionar els dígits de després del + i els guardar-los en una variable
    segundoval_entero = int(segundoval) #variable de tipus enter
    ret_val = primerval_entero + segundoval_entero; #sumar les variables
    return ret_val #retornar el valor


In [33]:
dz=df.loc[:,['LS','ST','RS','LW','LF','CF','RF','RW','LAM','CAM','RAM','LM','LCM','CM','RCM','RM','LWB','LDM','CDM','RDM','RWB',
            'LB','LCB','CB','RCB','RB']] #llista amb totes les columnes a les quals hem d'aplicar l'algoritme anterior
dz.head()

LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   
4  83+3  83+3  83+3  89+3  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   
4  82+3  82+3  82+3  89+3  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3   

    RCB    RB  
0  47+2  59+2  
1  53+3  61+3  
2  47+3  60+3  
3  66+3  73+3  
4  49+3  60+3

In [34]:
for i in dz:
    df[i] = df[i].apply(value_to_int) #bucle amb el qual apliquem l'algoritme a totes les columnes de dz
df.head()

Age  Overall  Potential        Value      Wage  International Reputation  \
0   31       94         94  110500000.0  565000.0                       5.0   
1   33       94         94   77000000.0  405000.0                       5.0   
2   26       92         93  118500000.0  290000.0                       5.0   
3   27       91         92  102000000.0  355000.0                       4.0   
4   27       91         91   93000000.0  340000.0                       4.0   

   Weak Foot  Skill Moves      Height     Weight  LS  ST  RS  LW  LF  CF  RF  \
0        4.0          4.0  173.735994  72.121927  90  90  90  94  95  95  95   
1        4.0          5.0  188.975994  83.008255  94  94  94  92  93  93  93   
2        5.0          5.0  179.831994  68.039554  87  87  87  92  92  92  92   
3        5.0          4.0  155.752795  69.853942  85  85  85  90  90  90  90   
4        4.0          4.0  176.783994  73.936315  86  86  86  92  91  91  91   

   RW  LAM  CAM  RAM  LM  LCM  CM  RCM  RM  LWB  LDM  CDM  RDM  RWB  LB  LCB  \
0  94   95   95   95  93   86  86   86  93   66   63   63   63   66  61   49   
1  92   91   91   91  91   84  84   84  91   68   64   64   64   68  64   56   
2  92   92   92   92  91   84  84   84  91   68   63   63   63   68  63   50   
3  90   91   91   91  91   90  90   90  91   80   80   80   80   80  76   69   
4  92   92   92   92  92   85  85   85  92   69   66   66   66   69  63   52   

   CB  RCB  RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  49   49  61      84.0       95.0             70.0          90.0     86.0   
1  56   56  64      84.0       94.0             89.0          81.0     87.0   
2  50   50  63      79.0       87.0             62.0          84.0     84.0   
3  69   69  76      93.0       82.0             55.0          92.0     82.0   
4  52   52  63      81.0       84.0             61.0          89.0     80.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       86.0   85.0        83.0         91.0         91.0          78.0   
4       95.0   83.0        79.0         83.0         94.0          94.0   

   SprintSpeed  Agility  Reactions  ...  nat_Venezuela  nat_Wales  nat_Zambia  \
0         86.0     91.0       95.0  ...              0          0           0   
1         91.0     87.0       96.0  ...              0          0           0   
2         90.0     96.0       94.0  ...              0          0           0   
3         76.0     79.0       91.0  ...              0          0           0   
4         88.0     95.0       90.0  ...              0          0           0   

   nat_Zimbabwe  pf_Left  pf_Right  bt_Akinfenwa  bt_C. Ronaldo  bt_Lean  \
0             0        1         0             0              0        0   
1             0        0         1             0              1        0   
2             0        0         1             0              0        0   
3             0        0         1             0              0        0   
4             0        0         1             0              0        0   

   bt_Messi  bt_Neymar  bt_Normal  bt_PLAYER_BODY_TYPE_25  bt_Shaqiri  \
0         1          0          0                       0           0   
1         0          0          0                       0           0   
2         0          1          0                       0           0   
3         0          0          1                       0           0   
4         0          0          1                       0           0   

   bt_Stocky  wr_High/ High  wr_High/ Low  wr_High/ Medium  wr_Low/ High  \
0          0              0             0                0             0   
1          0              0             1                0             0   
2          0              0  

## 7. Entrenar el dataset i realitzar la predicció

Eliminar la columna Value per a poder realitzar la predicció.

In [35]:
val = df.pop("Value")

In [36]:
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.33, random_state=42)

In [37]:
len(X_train)

10670

Entrenament del nostre data set basat en un model de regressió lineal.

In [38]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

In [39]:
preds = reg.predict(X_test)

In [40]:
preds[0]

557308.4934161701

In [41]:
y_test[0]

110500000.0

El resultat de la predicció:

In [42]:
r2_score(preds, y_test)

0.962053739744851

El resultat ha estat del 0.962053739744851.
Conclusió: el resultat és fiable ja que és un resultat bastant elevat (en un 96'21%).